File: Go/PollingSystem2-go/repositories/user_repository.go
package repositories

import (
	"fmt"
	"sync"

	"lld-practice/pollingsystem2-go/models"
)

type UserRepository struct {
	mu     sync.RWMutex
	users  []*models.User
	nextID int
}

func NewUserRepository() *UserRepository {
	return &UserRepository{users: []*models.User{}, nextID: 1}
}

func (r *UserRepository) Create(email string) (*models.User, error) {
	if email == "" {
		return nil, fmt.Errorf("invalid user email!")
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	for _, u := range r.users {
		if u.Email == email {
			return nil, fmt.Errorf("user already exists!")
		}
	}

	user, err := models.NewUser(r.nextID, email)
	if err != nil {
		return nil, err
	}
	r.nextID++
	r.users = append(r.users, user)
	return user, nil
}

File: JavaScript/PollingSystem2/PollingService/repositories/UserRepository.js
const { User } = require('../models/User');

/** In-memory stand-in for a user repository. */
class UserRepository {
  static users = [];
  static nextId = 1;

  static createId() {
    return this.nextId++;
  }

  static add(user) {
    if (!(user instanceof User)) {
      throw new Error('Invalid user!');
    }

    if (this.users.some((u) => u.id === user.id)) {
      throw new Error('User already exists!');
    }

    this.users.push(user);
  }

  static create(email = '') {
    if (!email) {
      throw new Error('Invalid user email!');
    }

    if (this.users.some((u) => u.email === email)) {
      throw new Error('User already exists!');
    }

    const user = new User(this.createId(), email);
    this.add(user);
    return user;
  }

  static getById(id) {
    return this.users.find((u) => u.id === id);
  }
}

module.exports = { UserRepository };
